### ISSUE TITLE
VictoriaLogs: Unwanted zero values returned by conditional count() stats by time buckets

### ISSUE THREAD

**Issue by p5i** [VictoriaMetrics/VictoriaMetrics#7162]
### Describe the bug

Hi.

When I do
`$ curl http://localhost:9428/select/logsql/query -d 'query= job:= messages AND _time:(1722919443899, 1722919446249) | stats by (_time:1ms) count() if(kernel) as kernel'`

I receive
```
{"_time":"2024-08-06T04:44:04.808Z","kernel":"1"}
{"_time":"2024-08-06T04:44:05.103Z","kernel":"0"}
{"_time":"2024-08-06T04:44:04.011Z","kernel":"2"}
{"_time":"2024-08-06T04:44:05.051Z","kernel":"2"}
{"_time":"2024-08-06T04:44:06.092Z","kernel":"2"}
{"_time":"2024-08-06T04:44:05.529Z","kernel":"4"}
{"_time":"2024-08-06T04:44:03.976Z","kernel":"0"}
{"_time":"2024-08-06T04:44:05.619Z","kernel":"0"}
```
I expect VictoriaLogs to return only rows with non-zero counts
```
{"_time":"2024-08-06T04:44:04.808Z","kernel":"1"}
{"_time":"2024-08-06T04:44:04.011Z","kernel":"2"}
{"_time":"2024-08-06T04:44:05.051Z","kernel":"2"}
{"_time":"2024-08-06T04:44:06.092Z","kernel":"2"}
{"_time":"2024-08-06T04:44:05.529Z","kernel":"4"}
```

Non-conditional count() works as expected on the same sample. Here the condition is moved to the filter
```
$ curl http://localhost:9428/select/logsql/query -d 'query= job:= messages AND kernel AND _time:[1722919443899, 1722919446249] | stats by (_time:1ms) count() as kernel'
{"_time":"2024-08-06T04:44:04.011Z","kernel":"2"}
{"_time":"2024-08-06T04:44:05.051Z","kernel":"2"}
{"_time":"2024-08-06T04:44:06.092Z","kernel":"2"}
{"_time":"2024-08-06T04:44:04.808Z","kernel":"1"}
{"_time":"2024-08-06T04:44:05.529Z","kernel":"4"}
```

### To Reproduce

This issue is easy to reproduce by generating timeseries with conditional count(). Almost exact queries can be reproduced on any Linux messages/syslog logs.

### Version

$ /opt/victoria-logs-prod --version
victoria-logs-20241001-115230-tags-v0.33.0-victorialogs-0-g234c81754e

### Logs

There is no records in logs.

### Additional information

When I tried to workaround this issue, I found problems with replace/replace_regex pipes. They are applied only to some rows randomly in this case
```
$ curl http://localhost:9428/select/logsql/query -d 'query= job:= messages AND _time:(1722919443899, 1722919446249) | stats by (_time:1ms) count() if(kernel) as kernel | replace(0, "") at kernel'
{"_time":"2024-08-06T04:44:06.092Z","kernel":"2"}
{"_time":"2024-08-06T04:44:04.808Z","kernel":"1"}
{"_time":"2024-08-06T04:44:05.103Z","kernel":""}
{"_time":"2024-08-06T04:44:03.976Z","kernel":"0"}
{"_time":"2024-08-06T04:44:05.529Z","kernel":"4"}
{"_time":"2024-08-06T04:44:05.619Z","kernel":""}
{"_time":"2024-08-06T04:44:04.011Z","kernel":"2"}
{"_time":"2024-08-06T04:44:05.051Z","kernel":"2"}
```
And with drop_empty_fields pipe
```
$ curl http://localhost:9428/select/logsql/query -d 'query= job:= messages AND _time:(1722919443899, 1722919446249) | stats by (_time:1ms) count() if(kernel) as kernel | replace(0, "") at kernel | drop_empty_fields'
{"_time":"2024-08-06T04:44:05.051Z","kernel":"2"}
{"_time":"2024-08-06T04:44:06.092Z","kernel":"2"}
{"_time":"2024-08-06T04:44:04.808Z","kernel":"1"}
{"_time":"2024-08-06T04:44:05.103Z"}
{"_time":"2024-08-06T04:44:05.529Z","kernel":"4"}
{"_time":"2024-08-06T04:44:03.976Z"}
{"_time":"2024-08-06T04:44:05.619Z","kernel":"0"}
{"_time":"2024-08-06T04:44:04.011Z","kernel":"2"}

```
According to docs
> | drop_empty_fields pipe drops fields with empty values. It also skips log entries with zero non-empty fields.

I think you can also drop entries with a single '_time' field.

**Comment by jiekun** [comment]
@p5i Hi. This should be an expected behaviour. 

You could place the filter before `stats`. How about the following LogsQL for your case?
```
kernel:(NOT "")| AND job:= messages AND _time:(1722919443899, 1722919446249) | stats by (_time:1ms) count() if(kernel) as kernel
```

**Comment by p5i** [comment]
@jiekun Hi. Thank you for your work.

This is a good point that the filter can be put before the 'stats'. But in this case there is no point in conditional stats. I can just use "stats count()" without "if".

My example was simplified. What I really tried to do with conditional filters is to form multiple time series with one query. As described here https://docs.victoriametrics.com/victorialogs/logsql/#stats-with-additional-filters
```

$ curl http://localhost:9428/select/logsql/query -d 'query= job:= messages AND _time:[1722919443899, 1722919446249]
| stats by (_time:1ms)
count() if(kernel) as kernel,
count() if(systemd) as systemd'

{"_time":"2024-08-06T04:44:05.619Z","kernel":"0","systemd":"0"}
{"_time":"2024-08-06T04:44:06.092Z","kernel":"2","systemd":"0"}
{"_time":"2024-08-06T04:44:04.011Z","kernel":"2","systemd":"0"}
{"_time":"2024-08-06T04:44:06.188Z","kernel":"0","systemd":"1"}
{"_time":"2024-08-06T04:44:06.184Z","kernel":"0","systemd":"0"}
{"_time":"2024-08-06T04:44:03.976Z","kernel":"0","systemd":"1"}
{"_time":"2024-08-06T04:44:04.808Z","kernel":"1","systemd":"0"}
{"_time":"2024-08-06T04:44:05.103Z","kernel":"0","systemd":"0"}
{"_time":"2024-08-06T04:44:05.529Z","kernel":"4","systemd":"0"}
{"_time":"2024-08-06T04:44:06.026Z","kernel":"0","systemd":"0"}
{"_time":"2024-08-06T04:44:05.051Z","kernel":"2","systemd":"0"}
```

With this output I can use Grafana's transformations to output kernel and systemd and any other facility as nice graph. The zeros make the graph look less demonstrative.
Of course this 'messages' example is made-up just to explain the point because it can be reproduced easily.

If I use filters before 'stats', as I really do now, than I need to execute multiple queries. Like this.
```
$ curl http://localhost:9428/select/logsql/query -d 'query= job:= messages AND kernel AND _time:[1722919443899, 1722919446249] | stats by (_time:1ms) count() as kernel'
{"_time":"2024-08-06T04:44:05.529Z","kernel":"4"}
{"_time":"2024-08-06T04:44:04.011Z","kernel":"2"}
{"_time":"2024-08-06T04:44:05.051Z","kernel":"2"}
{"_time":"2024-08-06T04:44:06.092Z","kernel":"2"}
{"_time":"2024-08-06T04:44:04.808Z","kernel":"1"}
```
```
$ curl http://localhost:9428/select/logsql/query -d 'query= job:= messages AND systemd AND _time:[1722919443899, 1722919446249] | stats by (_time:1ms) count() as systemd'
{"_time":"2024-08-06T04:44:03.976Z","systemd":"1"}
{"_time":"2024-08-06T04:44:06.188Z","systemd":"1"}
```
It works, but this is less effective and looks much more complex when creating many time series.

On the other hand I see no point in outputting zero counts in time series. And non-conditional stats count doesn't output zeros. This is why I think that it is not expected behavior.

Again, thank you very much for debugging this case.

**Comment by valyala** [comment]
Returning zero values from [conditional stats](https://docs.victoriametrics.com/victorialogs/logsql/#stats-with-additional-filters) is the expected behavior.

The [`stats by (...)`](https://docs.victoriametrics.com/victorialogs/logsql/#stats-by-fields) returns all the `by (...)` field value tuples seen in the matching logs. For example, it is expected that the `_time:5m | stats by (level) count()` query returns all the `level` field values seen in the logs over the last 5 minutes. If some [stats functions](https://docs.victoriametrics.com/victorialogs/logsql/#stats-pipe-functions) at `stats by (...)` pipe contain [`if (...)` filters](https://docs.victoriametrics.com/victorialogs/logsql/#stats-with-additional-filters), then this shouldn't change the set of returned `by (...)` field value tuples, even if some of the [stats functions](https://docs.victoriametrics.com/victorialogs/logsql/#stats-pipe-functions) return zero or an empty result because all the logs are filtered out with the attached `if (...)` filters. This simplifies understanding the logic behind `stats by (...)` pipe.

It is expected that some additional calculations are performed on `stats by(...)` results with the applied `if (...)` filters. For example, the following query calculates per-`host` share of logs with 200 status code over the last day:

```logsql
_time:1d
  | stats by (host)
    count() as hits_total,
    count() if (status:200) as hits_200
  | math hits_200 / hits_total as hits_200_share
```

Suppose there are no logs with `status=200` field value over the last day. In this case the `count() if (status:200)` returns `0`, so `hits_200 / hits_total` is properly calculated to `0`. If the `hits_200` field will be omitted in this case, then the `hits_200_share` will be set to `NaN`, which isn't expected by most users.

**Comment by valyala** [comment]
@p5i , it looks like you are trying to emulate [`/select/logsql/hits`](https://docs.victoriametrics.com/victorialogs/querying/#querying-hits-stats) endpoint via [`/select/logsql/query`](https://docs.victoriametrics.com/victorialogs/querying/#querying-logs) endpoint. Probably, it would be better using the `/select/logsql/hits` endpoint instead?

**Comment by valyala** [comment]
> When I tried to workaround this issue, I found problems with replace/replace_regex pipes. They are applied only to some rows randomly in this case

This is a separate bug, which must be fixed.

**Comment by valyala** [comment]
The bug with incorrect behavior for [`replace`](https://docs.victoriametrics.com/victorialogs/logsql/#replace-pipe) and [`replace_regexp`](https://docs.victoriametrics.com/victorialogs/logsql/#replace_regexp-pipe) pipes has been fixed in the commit 2e635a42d886fbd60d53ecfe69e43a233515bc5b . This commit will be included in the next release.

**Comment by valyala** [comment]
> I think you can also drop entries with a single '_time' field.

This will be unexpected behaviour. [`drop_empty_fields`](https://docs.victoriametrics.com/victorialogs/logsql/#drop_empty_fields-pipe) pipe must drop log lines only if they contain zero log fields, as documented [here](https://docs.victoriametrics.com/victorialogs/logsql/#drop_empty_fields-pipe).

**Comment by valyala** [comment]
FYI, the bug with `replace` and `replace_regexp` pipes has been fixed in [v0.40.0-victorialogs](https://docs.victoriametrics.com/victorialogs/changelog/). Closing the issue as addressed and answered.

**Comment by p5i** [comment]
@valyala, thanks for the great explanation of the logic behind the query language and for fixing the bug.

You are right that I can use the "/select/logsql/hits" endpoint for this query. I also use "/select/logsql/stats_query_range" in similar cases. I can't just find the standard way to use this endpoints in Grafana, so I use "/select/logsql/query".
